src/core/server/http/http_config.ts

```diff
@@ -25,10 +25,27 @@ import { SslConfig, sslSchema } from './ssl_config';
 
 const validBasePathRegex = /^\/.*[^\/]$/;
 const uuidRegexp = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-5][0-9a-f]{3}-[089ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
+const ipv4Regexp = /^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/;
+const ipv6Regexp = /^(?:[A-F0-9]{1,4}:){7}[A-F0-9]{1,4}$/i;
 
 const match = (regex: RegExp, errorMsg: string) => (str: string) =>
   regex.test(str) ? undefined : errorMsg;
 
+/** @internal */
+export const requestIdConfigSchema = schema.object({
+  allowFromAnyIp: schema.boolean({ defaultValue: true }),
+  ipAllowlist: schema.arrayOf(
+    schema.string({
+      validate: (value) => {
+        if (!ipv4Regexp.test(value) && !ipv6Regexp.test(value)) {
+          return 'must be a valid IPv4 or IPv6 address';
+        }
+      },
+    }),
+    { defaultValue: [] }
+  ),
+});
```

Comment:
Two things I don't love about this schema right now:

### Shape
I don't like the shape of this since it allows the user to easily misconfigure this. For instance if the user specified this in their config file:

```yaml
server.requestOpaqueId.ipAllowlist:
  - 1.1.1.1
```

The behavior would still allow any IP address because `allowFromAnyIp` is still defaulting to `true`. One option would be to add a custom validation to not allow this. The other would be to allow a more sensible shape like:

```ts
schema.object({
  allowFromIps: schema.oneOf([
    schema.literal('any'),
    schema.literal('none'),
    schema.arrayOf(schema.string(), { defaultValue: [] }),
  ], { defaultValue: 'any' }),
});
```

However it lead to strange validation errors if the IP address is not valid, for instance if configured like:

```yaml
server.requestOpaqueId.allowFromIps: [1234]
```

The (very confusing) error logged to the user would be:

```
{ Error: [allowFromIps]: types that failed validation:
- [allowFromIps.0]: expected value to equal [any]
- [allowFromIps.1]: expected value to equal [none]
- [allowFromIps.2.0]: expected value of type [string] but got [number]
```

### IP Validation

I've grabbed some regular expressions for validating IPv4 and IPv6 addresses from [this book](https://www.oreilly.com/library/view/regular-expressions-cookbook/9780596802837/ch07s16.html), however I'm paranoid that these may not be 100% correct. I'm not sure what is worse: no validation, or validation that erroneously rejects valid inputs (false negatives).

I lean towards removing this validation since invalid inputs would not severly affect the application. They simply would not match any incoming IP addresses and the X-Opaque-Id header on incoming requests would not be used.

Reply:
> The behavior would still allow any IP address because allowFromAnyIp is still defaulting to true

Why not set `allowFromAnyIp` to `false` if `ipAllowlist` is not empty in `HttpConfig` constructor?

> I've grabbed some regular expressions for validating IPv4 and IPv6 addresses from this book, however I'm paranoid that these may not be 100% correct.

The real question is what is the actual IP format used/returned by hapi `Request.info.remoteAddress`. Having other way to define IP in the config seems useless, if not dangerous. Else we need to adapt the very naive

```ts
this.options.requestOpaqueId.ipAllowlist.includes(this.socket.remoteAddress!)
        ? request.headers['x-opaque-id'] ?? uuid.v4()
        : uuid.v4();
```

check to properly convert to an homogeneous format before comparing.

Reply:
> Why not set allowFromAnyIp to false if ipAllowlist is not empty in HttpConfig constructor?

Or write a custom `validate` function? 
What is the case for `allowFromIps: none`? Can we treat an empty array as `none`? 

> IP Validation

Why not add `ip` to `@kbn/config-schema` that uses Joi under the hood? https://hapi.dev/module/joi/api/?v=17.1.1#stringipoptions That might resolve `The real question is what is the actual IP format used/returned by hapi`  as well.

Reply:
Added an `ip` type to config schema that uses Joi under the hood 👍